src/System.Reflection.Metadata.ILDasm/src/Decoder/ILDecoder.cs

```diff
@@ -0,0 +1,582 @@
+            }
+            return GetTypeInformation(mdReader, intOperand, provider);
+        }
+
+        private static string GetTypeInformation(MetadataReader mdReader, int intOperand, ILTypeProvider provider)
+        {
+            if(IsTypeReference(intOperand))
+            {
+                var refHandle = MetadataTokens.TypeReferenceHandle(intOperand);
+                return SignatureDecoder.DecodeType(refHandle, provider).ToString();
+            }
+            if (IsTypeSpecification(intOperand))
+            {
+                var typeHandle = MetadataTokens.TypeSpecificationHandle(intOperand);
+                return SignatureDecoder.DecodeType(typeHandle, provider).ToString();
+            }
+            var defHandle = MetadataTokens.TypeDefinitionHandle(intOperand);
+            return SignatureDecoder.DecodeType(defHandle, provider).ToString();
+        }
+
+        private static ILInstruction CreateSwitchInstruction(ref BlobReader ilReader, int expectedSize, int ilOffset, OpCode opCode)
+        {
+            var caseNumber = ilReader.ReadUInt32();
+            int[] jumps = new int[caseNumber];
+            for(int i = 0; i < caseNumber; i++)
+            {
+                jumps[i] = ilReader.ReadInt32();
+            }
+            int size = 4 + expectedSize;
+            size += (int)caseNumber * 4;
+            return new ILSwitchInstruction(opCode, ilOffset, jumps, (int)caseNumber, caseNumber, size);
+        }
+
+        private static string GetArgumentString(MetadataReader mdReader, int intOperand, out bool isPrintable)
+        {
+            if (IsUserString(intOperand))
+            {
+                UserStringHandle usrStr = MetadataTokens.UserStringHandle(intOperand);
+                var str = mdReader.GetUserString(usrStr);
+                str = ProcessAndNromalizeString(str, out isPrintable);
+                return str;
+            }
+            throw new NotImplementedException("Argument String");
+        }
+
+        private static string ProcessAndNromalizeString(string str, out bool isPrintable)
+        {
+            isPrintable = true;
+            StringBuilder sb = new StringBuilder();
+            sb.Append("bytearray (");
+            foreach (var c in str)
+            {
+                UnicodeCategory category = CharUnicodeInfo.GetUnicodeCategory(c);
+                if (isPrintable && (category == UnicodeCategory.Control || category == UnicodeCategory.OtherNotAssigned || category == UnicodeCategory.OtherSymbol || c == '"'))
+                {
+                    var _bytes = UnicodeEncoding.Unicode.GetBytes(str);
+                    foreach (var _byte in _bytes)
+                    {
+                        sb.Append(_byte.ToString("X2"));
+                        sb.Append(" ");
+                    }
+
+                    isPrintable = false;
+                    break;
+                }
+            }
+
+            sb.Append(")");
```

Comment:
It looks like this could be moved to after the subsequent isPrintable check, or into the loop above before isPrintable is set to false.  And if you do the latter, the creation of the StringBuilder could also move into the loop where it's filled in.  Then you could structure it like:
```C#
foreach (char c in str)
{
    UnicodeCategory category = ...;
    if (category == ...)
    {
        isPrintable = false;
        return SameLogicAsDiscussedEarlierForConvertingCharsToHexString(str);
    }
}
isPrintable = true;
return str;
```